### ISSUE TITLE
Core Lint error with binder swap

### ISSUE THREAD

**Issue by monoidal** [tweag/ghc#399]
The following code
```
    data Complex = !Float :+ !Float
    
    nan :: Float
    nan = 0+0
    
    mul :: Complex
    mul = nan :+ nan
```
when compiled with 91316cb654d54066005848053a24b841142a7699, shows linearity errors. The wrapper is

```
    Data.Complex.$W:+['Omega]
      = \ (dt_a16O['One] ::['One] Float)
          (dt_a16P['One] ::['One] Float) ->
          case dt_a16O['One] of (dt_X16R['One] ::['One] Float)
          { __DEFAULT ->
          case dt_a16P['One] of (dt_X16T['One] ::['One] Float)
          { __DEFAULT ->
          Data.Complex.:+['Omega] dt_X16R['One] dt_X16T['One]
          }
          }
```
when inlined in `mul`, the function `refineDefaultAlts` changes DEFAULT clauses to `F# (ipv_s1bS['One] :: Float#)`. This should be harmless; looks like Lint is wrong.

**Comment by aspiwack** [comment]
In fact, besides the error message attribution the issue to a random variable, this issue seems to be a case of faulty binder-swap. The term which doesn't pass the linter is:

```haskell
mul :: Complex
[LclIdX,
 Unf=Unf{Src=<vanilla>, TopLvl=True, Value=False, ConLike=False,
         WorkFree=True, Expandable=False, Guidance=IF_ARGS [] 20 30}]
mul = case nan of dt_XZu { F# ipv_s1bS -> :+ dt_XZu dt_XZu }
```

So `dt_XZu` is used non-linearly.

**Comment by monoidal** [comment]
Quick fix implemented in 5a3c3d3c64f1b818f853c8eac3f9a773d2470012.

**Comment by aspiwack** [comment]
This is good. As long as you add a comment to the effect that binder-swap is not generally legal with linear binders.
